src/plots/cartesian/constraints.js

```diff
@@ -143,15 +144,17 @@ exports.enforce = function enforceAxisConstraints(gd) {
                         var newVal;
                         var k;
 
-                        for(k = 0; k < ax._min.length; k++) {
-                            newVal = ax._min[k].val - getPad(ax._min[k]) / m;
+                        var minArray = concatExtremes(gd, ax, 'min');
```

Comment:
Hmm, so looks like we can call `concatExtremes` twice with the same results? Could we cache? Like, clear it maybe during `supplyDefaults`, then in `concatExtremes` first look for the cached value before calculating it and setting the cache?

Reply:
Some improvement in 66df51e8cf9d99e69f3968eb5e73929c55c233a2. `concatExtremes` is still called twice on graphs with constrained axes, so would do better if we want.